### PR TITLE
Add documentation for JAXJob

### PR DESCRIPTION
**Pull request by sandipanpanda**
Ref: https://github.com/kubeflow/training-operator/issues/1619 https://github.com/kubeflow/training-operator/issues/2145

### CODE REVIEW

**Review by Arhell — COMMENTED** [review]
/ok-to-test

**Review by andreyvelich — COMMENTED** [review]
Thank you for doing this @sandipanpanda!
Let's merge it once we complete the Jax implementation in Training Operator.
/hold
/assign @kubeflow/wg-training-leads @StefanoFioravanzo @hbelmiro for review

**Review by hbelmiro — COMMENTED** [review]
/lgtm

**Review by hbelmiro — COMMENTED** [review]
/lgtm

File: content/en/docs/components/training/user-guides/jax.md [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
++++
+
+This page describes `JAXJob` for training a machine learning model with [JAX](https://jax.readthedocs.io/en/latest/).
+
+The `JAXJob` is a Kubernetes
+[custom resource](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/)
+to run JAX training jobs on Kubernetes. The Kubeflow implementation of
+the `JAXJob` is in the [`training-operator`](https://github.com/kubeflow/training-operator).
+
+The current custom resource for JAX has been tested to run multiple processes on CPUs using [gloo](https://github.com/facebookincubator/gloo) for communication between CPUs.
```
Comment (andreyvelich) [inline content/en/docs/components/training/user-guides/jax.md:14]: @kubeflow/wg-training-leads @sandipanpanda Do we want to mention that we are looking for user feedback to run JAXJob on TPUs ?

File: content/en/docs/components/training/user-guides/jax.md [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+## Creating a JAX training job
+
+You can create a training job by defining a `JAXJob` config file. See the manifests for the [simple JAXJob example](https://github.com/kubeflow/training-operator/blob/master/examples/jax/cpu-demo/demo.yaml).
+You may change the config file based on your requirements.
```
Comment (andreyvelich) [inline content/en/docs/components/training/user-guides/jax.md:19]: What do you mean by changing the config file here ?

File: content/en/docs/components/training/user-guides/jax.md [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+kubectl create -f https://raw.githubusercontent.com/kubeflow/training-operator/refs/heads/master/examples/jax/cpu-demo/demo.yaml
+```
+
+You should now be able to see the created pods matching the specified number of replicas.
+
+```
+kubectl get pods -n kubeflow -l training.kubeflow.org/job-name=jaxjob-simple
+```
+
+Training takes 5-10 minutes on a CPU cluster. Logs can be inspected to see its training progress.
```
Comment (andreyvelich) [inline content/en/docs/components/training/user-guides/jax.md:33]: What do you mean by training if we just compute all-reduce sum across all JAX processes: https://github.com/kubeflow/training-operator/blob/master/examples/jax/cpu-demo/train.py#L39C15-L39C19 ?

File: content/en/docs/components/training/user-guides/jax.md [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+```
+PODNAME=$(kubectl get pods -l training.kubeflow.org/job-name=jaxjob-simple,training.kubeflow.org/replica-type=worker,training.kubeflow.org/replica-index=0 -o name -n kubeflow)
+kubectl logs -f ${PODNAME} -n kubeflow
```
Comment (andreyvelich) [inline content/en/docs/components/training/user-guides/jax.md:37]: Please can you add the output of the logs here.

**Review by andreyvelich — COMMENTED** [review]
Thank you for adding this @sandipanpanda!
I left a few comments.

**Review by terrytangyuan — APPROVED** [review]
Thank you! 

/lgtm

**Comment by andreyvelich** [comment]
@sandipanpanda Did you get a chance to review the remaining comments ?

File: content/en/docs/components/training/user-guides/jax.md (hunk 1 above)
Comment (tenzen-y) [reply content/en/docs/components/training/user-guides/jax.md:14]: User feedback would be a great idea. But, even if we get any feedback for the TPU, I'm wondering if we can not implement TPU support in the upstream training-operator because we do not have any verification infrastructure, right?

File: content/en/docs/components/training/overview.md [hunk 5]
```diff
@@ -10,7 +10,7 @@ weight = 10
 
 The Training Operator is a Kubernetes-native project for fine-tuning and scalable
 distributed training of machine learning (ML) models created with different ML frameworks such as
-PyTorch, TensorFlow, XGBoost, and others.
+PyTorch, TensorFlow, XGBoost, [JAX](https://jax.readthedocs.io/en/latest/), and others.
```
Comment (tenzen-y) [inline content/en/docs/components/training/overview.md:13]:
```suggestion
PyTorch, TensorFlow, XGBoost, JAX, and others.
```

For consistently across all supported frameworks.

File: content/en/docs/components/training/user-guides/jax.md [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+This page describes `JAXJob` for training a machine learning model with [JAX](https://jax.readthedocs.io/en/latest/).
+
+The `JAXJob` is a Kubernetes
+[custom resource](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/)
+to run JAX training jobs on Kubernetes. The Kubeflow implementation of
+the `JAXJob` is in the [`training-operator`](https://github.com/kubeflow/training-operator).
+
+The current custom resource for JAX has been tested to run multiple processes on CPUs using [gloo](https://github.com/facebookincubator/gloo) for communication between CPUs.
+
```
Comment (tenzen-y) [inline content/en/docs/components/training/user-guides/jax.md:15]: Additionally, could you mention that the worker with replica 0 is recognized as a JAX coordinator?
IIUC, we do not deploy the dedicated JAX Coordinator replicas, right?

File: content/en/docs/components/training/user-guides/jax.md (hunk 1 above)
Comment (andreyvelich) [reply content/en/docs/components/training/user-guides/jax.md:14]: I feel that it is fine, since we don't have infra for GPUs today as well.
E.g. we say that you can run those examples on GPUs, but we don't validate them in our CI.

File: content/en/docs/components/training/user-guides/jax.md (hunk 2 above)
Comment (sandipanpanda) [reply content/en/docs/components/training/user-guides/jax.md:19]: I meant the Job config file, using the wording in the existing [user guides](https://www.kubeflow.org/docs/components/training/user-guides/pytorch/#creating-a-pytorch-training-job). I'll make that clear.

File: content/en/docs/components/training/user-guides/jax.md (hunk 3 above)
Comment (sandipanpanda) [reply content/en/docs/components/training/user-guides/jax.md:33]: I used it in the context of using JAXJob for distributed training. I'll reword it to computation as it makes for sense.

**Review by andreyvelich — COMMENTED** [review]
Looks great, thanks for the update @sandipanpanda 🎉 
/lgtm
/hold for others
/assign @kubeflow/wg-training-leads

File: content/en/docs/components/training/user-guides/jax.md (hunk 1 above)
Comment (tenzen-y) [reply content/en/docs/components/training/user-guides/jax.md]: > I feel that it is fine for now, since we don't have infra for GPUs today as well.
E.g. we say that you can run those examples on GPUs, but we don't validate them in our CI.

Yeah, that's true, but GPUs are mostly generic devices, and there are many developers who can access them. So, we can improve GPU utilization mechanism based on the GPUs.

OTOH, the TPU is only available in Google Cloud, and there are developers who can access it less than GPU one.
So, my concern is the TPU-specific mechanisms will be abandoned and will not work soon.

**Review by tenzen-y — COMMENTED** [review]
/lgtm

**Comment by andreyvelich** [comment]
Thanks for this @sandipanpanda!
/hold cancel
